src/methods/join.js

```diff
@@ -445,10 +458,31 @@ export class JoinImplementation {
     let empty_effects = construct_empty_effects(realm);
     let v = realm.intrinsics.empty;
     if (c.consequent instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, v, empty_effects, c.alternate, c.alternateEffects);
+      let negation = AbstractValue.createFromUnaryOp(realm, "!", c.joinCondition);
+      invariant(negation instanceof AbstractValue);
+      let pathCondition = [negation];
+      let pnc = new PossiblyNormalCompletion(
+        v,
+        c.joinCondition,
+        v,
+        empty_effects,
+        c.alternate,
+        c.alternateEffects,
+        pathCondition,
+        []
+      );
       return [c.consequentEffects, pnc];
     } else if (c.alternate instanceof ReturnCompletion) {
-      let pnc = new PossiblyNormalCompletion(v, c.joinCondition, c.consequent, c.consequentEffects, v, empty_effects);
+      let pnc = new PossiblyNormalCompletion(
+        v,
+        c.joinCondition,
+        c.consequent,
+        c.consequentEffects,
+        v,
+        empty_effects,
+        [c.joinCondition],
```

Comment:
You need to account for the possibility that c.consequent may be a PossiblyNormalCompletion.